### PR TITLE
Adding clean as a default task

### PR DESCRIPTION
**Pull request by msoule**

### CODE REVIEW

**Comment by msoule** [comment]
First time submitting a PR for this project. I have a few questions:

1. Unit tests for MainModule? I didn't see any.
2. Is it acceptable that I used `ammonite.ops.%%`?
3. Can `ammonite.ops.%%` throw an exception?
4. I opted for multi clean targets because it gave me an elegant way to do clean all (by passing no arguments). Is there a use case for nested clean or transitive dependency clean?
5. What are the styling standards for this project?

**Comment by msoule** [comment]
Also, one quirk is that clean does not clean up after itself. So after running clean you are left with `out/clean`. Not sure if there is anything I can do about that.

**Comment by msoule** [comment]
Fixes #227

File: main/src/mill/main/MainModule.scala [hunk 1]
```diff
@@ -173,4 +175,17 @@ trait MainModule extends mill.Module{
...
     }
   }
+
+  /**
+    * Deletes the given targets from the out directory. Providing no targets will clean everything.
+    */
+  def clean(targets: String*) = mill.T.command {
+    val result = targets match {
+      case Nil =>
+        List(ammonite.ops.%%('rm, "-rf", OutDir)(ammonite.ops.pwd).out.string.trim())
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: probably should use `ammonite.ops.rm` here

File: main/src/mill/main/MainModule.scala [hunk 2]
```diff
@@ -173,4 +175,17 @@ trait MainModule extends mill.Module{
...
+      case _ =>
+        targets.map(cleanTarget => ammonite.ops.%%('rm, "-rf", s"$OutDir/$cleanTarget")(ammonite.ops.pwd).out.string.trim())
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: This won't work for targets like `foo.bar.baz`

File: main/src/mill/main/MainModule.scala (hunk 1 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Aw okay didn't even know that was a thing.

File: main/src/mill/main/MainModule.scala (hunk 2 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Yeah I wanted to ask about this first to confirm that it is desirable. Will mark as WIP until I've added this.

File: main/src/mill/main/MainModule.scala (hunk 1 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Good point. I did note that other functions in this file do the following:

```
    val resolved = RunScript.resolveTasks(
      mill.main.ResolveMetadata, evaluator, targets, multiSelect = true
    )
```
And then I extract the names from `resolved`. Would that be sufficient? I will play with it later.

Comment (francisdb) [reply main/src/mill/main/MainModule.scala]: would it not be enough to validate that what you pass to rm -rf contains `out`?

Comment (francisdb) [reply main/src/mill/main/MainModule.scala]: And maybe that the parent folder contains build.sc?

Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Yes you're right. I guess I'm wondering if there is also value is validating that the passed targets are indeed valid targets.

Comment (francisdb) [reply main/src/mill/main/MainModule.scala]: You probably want to convert the path to a absolute path first before applying tests.

File: main/src/mill/main/MainModule.scala [hunk 3]
```diff
@@ -173,4 +179,34 @@ trait MainModule extends mill.Module{
       }
     }
   }
+
+  /**
+    * Deletes the given targets from the out directory. Providing no targets will clean everything.
+    */
+  def clean(targets: String*) = mill.T.command {
+    val result = Try {
+      if (ammonite.ops.ls(ammonite.ops.pwd).contains(ammonite.ops.pwd / BuildFile)) {
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: This can just be `exists(pwd / "build.sc")`

File: main/src/mill/main/MainModule.scala [hunk 4]
```diff
@@ -173,4 +179,34 @@ trait MainModule extends mill.Module{
...
+        targets match {
+          case Nil =>
+            remove(List(ammonite.ops.pwd / OutDir))
+          case _ =>
+            remove(targets
+              .map(_.replace(".", "/"))
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: You should use `resolveDestPaths` (https://github.com/lihaoyi/mill/blob/master/core/src/mill/eval/Evaluator.scala#L342) to figure out what targets to clean, to ensure it stays in sync with any future changes

File: main/src/mill/main/MainModule.scala [hunk 5]
```diff
@@ -173,4 +179,34 @@ trait MainModule extends mill.Module{
       }
     }
   }
+
+  /**
+    * Deletes the given targets from the out directory. Providing no targets will clean everything.
+    */
+  def clean(targets: String*) = mill.T.command {
+    val result = Try {
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: No idea why you are wrapping this in a `Try`; you just need to return a simple `Result` object

File: main/src/mill/main/MainModule.scala [hunk 6]
```diff
@@ -173,4 +179,34 @@ trait MainModule extends mill.Module{
...
+      }
+    }
+    result match {
+      case Success(r) => r
+      case Failure(error) => Result.Exception(error, new Result.OuterStack(error.getStackTrace))
+    }
+  }
+
+  private def remove(paths: List[Path]): Result[String] = {
+    paths.filter(ammonite.ops.exists).foreach(ammonite.ops.rm)
```
Comment (lihaoyi) [inline main/src/mill/main/MainModule.scala]: You don't need the `filter`; `rm` already just ignores a path if there' nothing to remove

**Comment by lihaoyi** [comment]
This should have at least one unit test somewhere before merging

**Comment by msoule** [comment]
@lihaoyi Where are the unit tests for the other functions in this file? Are there any?

File: main/src/mill/main/MainModule.scala (hunk 5 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: The ammonite functions call things which can throw exceptions. Is it acceptable to let those exceptions be thrown?

File: main/src/mill/main/MainModule.scala (hunk 6 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Oh nice.

File: main/src/mill/main/MainModule.scala (hunk 3 above)
Comment (msoule) [reply main/src/mill/main/MainModule.scala]: Will do.

**Comment by msoule** [comment]
Also @lihaoyi Can you explain the build failures I'm seeing. The same test passes locally.

**Comment by lihaoyi** [comment]
Closing due to inactivity

**Comment by msoule** [comment]
@lihaoyi Can you answer some of the questions I have above?

**Pull request closed without merging**